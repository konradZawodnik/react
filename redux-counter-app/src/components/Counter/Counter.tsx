import { useEffect, useState } from "react";
import { createStore } from "../../redux/store/store";
import { counterReducer } from "../../redux/reducer/reducer";
import { loggerMiddleware } from "@/redux/middleware/middleware";

const initialState = { count: 0 };
const store = createStore(counterReducer, initialState, [loggerMiddleware]);

const Counter: React.FC = () => {
  const [count, setCount] = useState(store.getState().count);

  useEffect(() => {
    const unsubscribe = store.subscribe(() => {
      setCount(store.getState().count);
    });

    return () => {
      unsubscribe();
    };
  }, []);

  const increment = () => store.dispatch({ type: "INCREMENT" });
  const decrement = () => store.dispatch({ type: "DECREMENT" });

  return (
    <div className="flex flex-col font-bold h-28 justify-center items-center">
      <h6 className="font-bold pb-4 text-base">
        Counter using simple tool to managing of app state
      </h6>
      <div className="flex-row pb-4">
        <button
          className="rounded-md pr-4 bg-slate-800 py-2 px-4 border border-transparent text-center text-sm text-white transition-all shadow-md hover:shadow-lg focus:bg-slate-700 focus:shadow-none active:bg-slate-700 hover:bg-slate-700 active:shadow-none disabled:pointer-events-none disabled:opacity-50 disabled:shadow-none ml-2"
          onClick={increment}
          type="button"
        >
          Increment
        </button>
        <button
          className="rounded-md bg-slate-800 py-2 px-4 border border-transparent text-center text-sm text-white transition-all shadow-md hover:shadow-lg focus:bg-slate-700 focus:shadow-none active:bg-slate-700 hover:bg-slate-700 active:shadow-none disabled:pointer-events-none disabled:opacity-50 disabled:shadow-none ml-2"
          onClick={decrement}
          type="button"
        >
          Decrement
        </button>
      </div>
      <h1 className="text-base" data-testid="count-text">
        Count: {count}
      </h1>
    </div>
  );
};

export default Counter;
