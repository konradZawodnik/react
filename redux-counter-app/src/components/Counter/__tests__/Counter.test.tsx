import { render, screen, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import Counter from "../Counter";

describe("Counter", () => {
  beforeEach(() => {
    render(<Counter />);
  });

  it("should match snapshot", () => {
    const counter = render(<Counter />);
    expect(counter).toMatchSnapshot();
  });

  it("renders initial count", () => {
    expect(screen.getByTestId("count-text")).toHaveTextContent("Count: 0");
  });

  it("increments the count when the Increment button is clicked", () => {
    const incrementButton = screen.getByText(/increment/i);
    fireEvent.click(incrementButton);

    expect(screen.getByTestId("count-text")).toHaveTextContent("Count: 1");
  });

  it("decrements the count when the Decrement button is clicked", () => {
    const decrementButton = screen.getByText(/decrement/i);
    fireEvent.click(decrementButton);

    expect(screen.getByTestId("count-text")).toHaveTextContent("Count: 0");
  });

  it("handles multiple increments and decrements correctly", () => {
    const incrementButton = screen.getByText(/increment/i);
    const decrementButton = screen.getByText(/decrement/i);
    fireEvent.click(incrementButton);
    fireEvent.click(incrementButton);
    fireEvent.click(decrementButton);

    expect(screen.getByTestId("count-text")).toHaveTextContent("Count: 1");
  });
});
