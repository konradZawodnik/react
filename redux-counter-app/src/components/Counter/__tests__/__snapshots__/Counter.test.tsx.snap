// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex flex-col font-bold h-28 justify-center items-center"
      >
        <h6
          class="font-bold pb-4 text-base"
        >
          Counter using simple tool to managing of app state
        </h6>
        <div
          class="flex-row pb-4"
        >
          <button
            class="rounded-md pr-4 bg-slate-800 py-2 px-4 border border-transparent text-center text-sm text-white transition-all shadow-md hover:shadow-lg focus:bg-slate-700 focus:shadow-none active:bg-slate-700 hover:bg-slate-700 active:shadow-none disabled:pointer-events-none disabled:opacity-50 disabled:shadow-none ml-2"
            type="button"
          >
            Increment
          </button>
          <button
            class="rounded-md bg-slate-800 py-2 px-4 border border-transparent text-center text-sm text-white transition-all shadow-md hover:shadow-lg focus:bg-slate-700 focus:shadow-none active:bg-slate-700 hover:bg-slate-700 active:shadow-none disabled:pointer-events-none disabled:opacity-50 disabled:shadow-none ml-2"
            type="button"
          >
            Decrement
          </button>
        </div>
        <h1
          class="text-base"
          data-testid="count-text"
        >
          Count: 
          0
        </h1>
      </div>
    </div>
    <div>
      <div
        class="flex flex-col font-bold h-28 justify-center items-center"
      >
        <h6
          class="font-bold pb-4 text-base"
        >
          Counter using simple tool to managing of app state
        </h6>
        <div
          class="flex-row pb-4"
        >
          <button
            class="rounded-md pr-4 bg-slate-800 py-2 px-4 border border-transparent text-center text-sm text-white transition-all shadow-md hover:shadow-lg focus:bg-slate-700 focus:shadow-none active:bg-slate-700 hover:bg-slate-700 active:shadow-none disabled:pointer-events-none disabled:opacity-50 disabled:shadow-none ml-2"
            type="button"
          >
            Increment
          </button>
          <button
            class="rounded-md bg-slate-800 py-2 px-4 border border-transparent text-center text-sm text-white transition-all shadow-md hover:shadow-lg focus:bg-slate-700 focus:shadow-none active:bg-slate-700 hover:bg-slate-700 active:shadow-none disabled:pointer-events-none disabled:opacity-50 disabled:shadow-none ml-2"
            type="button"
          >
            Decrement
          </button>
        </div>
        <h1
          class="text-base"
          data-testid="count-text"
        >
          Count: 
          0
        </h1>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex flex-col font-bold h-28 justify-center items-center"
    >
      <h6
        class="font-bold pb-4 text-base"
      >
        Counter using simple tool to managing of app state
      </h6>
      <div
        class="flex-row pb-4"
      >
        <button
          class="rounded-md pr-4 bg-slate-800 py-2 px-4 border border-transparent text-center text-sm text-white transition-all shadow-md hover:shadow-lg focus:bg-slate-700 focus:shadow-none active:bg-slate-700 hover:bg-slate-700 active:shadow-none disabled:pointer-events-none disabled:opacity-50 disabled:shadow-none ml-2"
          type="button"
        >
          Increment
        </button>
        <button
          class="rounded-md bg-slate-800 py-2 px-4 border border-transparent text-center text-sm text-white transition-all shadow-md hover:shadow-lg focus:bg-slate-700 focus:shadow-none active:bg-slate-700 hover:bg-slate-700 active:shadow-none disabled:pointer-events-none disabled:opacity-50 disabled:shadow-none ml-2"
          type="button"
        >
          Decrement
        </button>
      </div>
      <h1
        class="text-base"
        data-testid="count-text"
      >
        Count: 
        0
      </h1>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
